Hyperboloidal

dohertycw opened the pull request:
Addition of Hyperboloidal Method for circular orbits. Access this using ```Method -> "Hyperboloidal"``` in ```ReggeWheelerPointParticleMode[ ]```.

dohertycw committed:
Hyperboloidal init

dohertycw committed:
Changes to ReggeWheeler.m

dohertycw committed:
Added preliminary pre-14.0 compatible version

Commit:
Make Keys work on a ReggeWheelerMode object

Commit:
Make ReggeWheelerHyperboloidal output a ReggeWheelerMode, and make the fluxes work correctly by introducing a Method key in the ReggeWheelerMode association

Commit:
Integrate the new hyperboloidal code into the current API. Now you access it via ReggeWheelerPointParticleMode with a Method->"Hyperboloidal". You can also pass options to the hyperboloidal method via, e.g., Method -> {"Hyperboloidal", "GridPoints" -> 40}

dohertycw committed:
Removed unnecessary If[] statement in ReggeWheelerMode.m at line 72

dohertycw committed:
Numerous changes. Hyperboloidal error messages migrated to ReggeWheelerMode package, removal of unnecessary xi function in ReggeWheelerHyperboloidal, Hyperboloidal method "Radial Function" (name will be changed) key correctly returns solution in t-slicing, removal of unnecessary construction of ReggeWheelerHyperboloidalMode object and related associations and replaced with a call to the existing ReggeWheelerMode object, "WorkingPrecision" option for ReggeWheelerHyperboloidal removed, working precision now based on user specified orbital radius.

nielsw2 commented:
You can also change the number of grid points using the option:

```Mathematica
mode = ReggeWheelerPointParticleMode[2, 2, 2, 0, orbit, Method -> {"Hyperboloidal", "GridPoints" -> 32}]
```

Things that still need to be done:
- Add analytic mesh refinement
- Static modes are implemented (we change this to provide an error message for now).
- Add Documentation (we need to update the documentation for the entire package to the new system)
- Get the output to agree with the rest of the package (this is changing as @joshbmat is changing to output RW and Zerilli-Moncrief  as part of the eccentric orbit work).

nielsw2 commented:
Also, unit tests, but I suspect we need those across the package.

nielsw2 commented:
And we also need to copy the code from the Teukolsky package that allows you to evaluate a PointParticleMode at a given radius (and allows on the worldline/within the libration region if the ExtendedHomogeneous option is passed).

dohertycw committed:
Fixed error messages not correctly returning ```$Failed``` for ```Method -> Hyperboloidal```

dohertycw committed:
Added Analytic Mesh Refinement technique for more accurate horizon domain field solutions.

dohertycw committed:
Replaced usage of Mathematica's experimental ```BlockDiagonalMatrix``` function with ```ArrayFlatten[]``` (more efficient).

dohertycw committed:
Replaced usage of Mathematica's  function with

dohertycw committed:
Merge branch 'hyperboloidal' of https://github.com/BlackHolePerturbationToolkit/ReggeWheeler into hyperboloidal

dohertycw committed:
Added automatic check for minimum necessary working precision, with an interpolator (based on previous data) to automatically raise working precision to the minimum necessary.

dohertycw committed:
Documentation for the ReggeWheeler package has been updated to be inline with that of the Teukolsky package

dohertycw committed:
Removed differentiation using Mathematica's D[] function for the derivative(s) of our ansatz. Differentiation is now carried out using Chebyshev differentiation matrices.

dohertycw committed:
Small optimisations.

dohertycw committed:
Accuracy and efficiency improvements.

A third domain now spans from the particle to the horizon. This has shown marked improvements in agreement with the existing MST package. LARHS.Inverse[Mat] has been replaced with LinearSolve[Transpose[Mat],LARHS] for efficiency improvements.

dohertycw committed:
Fixed typo in odd sector alpha expressions.

dohertycw committed:
Addition of CITATION.cff and small changes to Hyperboloidal.wl

dohertycw committed:
Changes to Hyperboloidal.wl to ensure no changes needed to be made for
Energy Flux calculations in ReggeWheelerMode.m

Commit:
Return the eigenvalue in the mode when using the Hyperboloidal method

dohertycw committed:
Changes to Hyperboloidal.wl to ensure consistency with all other
methods when calculating fluxes and corrections to amplitude
caclulations. Changes to ReggeWheelerMode.wl to remove specific
expressions for calculating fluxes in the hyperboloidal case (all flux
calculations now use the same expressions).

Commit:
Quick fix for the amplitudes (omega factor defined after it was used)

nielsw2 commented:
I was running some tests on the latest code. I get an bunch of errors with, e.g., ```ReggeWheelerPointParticleMode[2, 2, 2, 0, orbit, 
 Method -> "Hyperboloidal"]``` if the orbit has a radius less than 10. The errors related to an interpolating function.

dohertycw committed:
Fixed issue with ReggeWheelerHyperboloidal not working for r < 10M.

dohertycw committed:
Merge branch 'hyperboloidal' of https://github.com/BlackHolePerturbationToolkit/ReggeWheeler into hyperboloidal

dohertycw committed:
Fixed package not working for r < 10M

nielsw2 commented:
It looks like the latest push does not fix the issue with r0 < 10 splitting out a warning about interpolation. Looking at the code I think this is coming from the WorkingPrecision Interpolation. I couldn't understand how that works so not sure how to fix it.

Also, how well tested is the MantissaExponent fix? It looks pretty hacky to just have if r<10 then it equals 1/2.

Despite the above the code does seem to work for r 0< 10 and the flux and amplitudes are correct when I tested for a range of r0 values and a few lm-modes.

nielsw2 commented:
Also, do we need the InterpolatorData.dat in the repository? It looks like the data is now in the Hyperboloidal.wl file.

dohertycw committed:
Removed InterpolatorData.dat as the data is in Hyperboloidal.wl. Small
cleanup in Hyperboloidal.wl and ReggeWheelerMode.m

dohertycw committed:
Fixes to the Working Precision Interpolator.
For r values < 10M, the precision provided by 10M being passed through
the Working Precision Interpolator is sufficient.

nielsw2 commented:
After the last push all the tests are passing. I will merge this branch now.
